packages/framework/undo-redo/src/mapHandler.ts

```diff
@@ -37,10 +51,23 @@ export class SharedMapRevertible implements IRevertible {
     ) { }
 
     public revert() {
-        this.map.set(this.changed.key, this.changed.previousValue);
+        if (this.changed.previousValue === undefined) {
+            this.map.delete(this.changed.key);
+        } else {
+            this.map.set(this.changed.key, this.changed.previousValue);
+        }
     }
 
     public discard() {
         return;
     }
+
+    /**
+     * Compare change to be rolled back against this revertible
+     */
+    public matchRollback?(key: string, target: ISharedMap): boolean {
```

Comment:
i think i'd rather leave undo redo out of this for now. there are interesting problems to solve with the intersection between rollback and undo redo, but i think we should design those.

specifically, i try not to enforce policy, and only define general mechanisms that can be used to implement whatever policy is desired by the user. I'm not actually sure what the state machine around order sequentially is. i know on receiving a batch we have batch begin and end events. if we don't have that locally we probably want that. then we'd also want a rollback event. having these would let an application open a new operation on the undo stack as batch being, and throw it away on rollback if that is the desired behavior. another option could be to treat both the initial batach begin and the rollback and undoable actions. The scenario i'm thinking of is some kind of drafting flow, which may not be possible today with the sync only requirement, but potentially something we could want eventually.